src/any_llm/utils/exception_handler.py

```
@@ -38,6 +38,108 @@
 
 
+# Message/type-name patterns tried in order; the first match wins, so more
+# specific patterns must stay ahead of broader ones (for example the API-key
+# patterns before the bare "invalid"). Anything unmatched becomes a ProviderError.
+_ERROR_PATTERNS: tuple[tuple[str, type[AnyLLMError]], ...] = (
+    (r"ratelimit|rate_limit|too many requests|rate limit|quota exceeded", RateLimitError),
+    (
+        r"auth|permission|invalid api key|invalid key|unauthorized|authentication|"
+        r"permission denied|access denied|forbidden|invalid_api_key|api key not found|api key invalid|"
+        r"api key not valid|incorrect api key|not valid.*api key|"
+        r"api key.*invalid|invalid.*api key",
```

Ruff (0.16.0)

[warning] Unparenthesized implicit string concatenation in collection

Did you forget a comma?

(ISC004)